lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,205 @@
+
+  void Post(const parser::AllocateStmt &) {
+    if (ObjectIsCoarray()) {
+      messages_.Say(charBlock_,
+          parser::MessageFormattedText{
+              "ALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US});
+    }
+  }
+  void Post(const parser::DeallocateStmt &) {
+    if (ObjectIsCoarray()) {
+      messages_.Say(charBlock_,
+          parser::MessageFormattedText{
+              "DEALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US});
+    }
+    // C1140: deallocation of polymorphic objects
+  }
+  template<typename T> void Post(const parser::Statement<T> &) {
+    if (EndTDeallocatesCoarray()) {
+      messages_.Say(charBlock_,
+          parser::MessageFormattedText{
+              "implicit deallocation of coarray not allowed"
+              " in DO CONCURRENT"_err_en_US});
+    }
+  }
+  void Post(const parser::CallStmt &) {
+    // C1137: call move_alloc with coarray arguments
+    // C1139: call to impure procedure
+  }
+  // C1141
+  void Post(const parser::ProcedureDesignator &procedureDesignator) {
+    if (auto *name = std::get_if<parser::Name>(&procedureDesignator.u)) {
+      auto upperName{parser::ToUpperCaseLetters(name->ToString())};
+      if (upperName == "IEEE_GET_FLAG"s) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{
+                "IEEE_GET_FLAG not allowed in DO CONCURRENT"_err_en_US});
+      } else if (upperName == "IEEE_SET_HALTING_MODE"s) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{"IEEE_SET_HALTING_MODE not allowed"
+                                         " in DO CONCURRENT"_err_en_US});
+      } else if (upperName == "IEEE_GET_HALTING_MODE"s) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{"IEEE_GET_HALTING_MODE not allowed"
+                                         " in DO CONCURRENT"_err_en_US});
+      }
+    }
+  }
+
+  // C1138: extended ranges in DOs should be errors, not warnings
+
+  // 11.1.7.5
+  void Post(const parser::IoControlSpec &ioControlSpec) {
+    if (auto *charExpr =
+            std::get_if<parser::IoControlSpec::CharExpr>(&ioControlSpec.u)) {
+      if (std::get<parser::IoControlSpec::CharExpr::Kind>(charExpr->t) ==
+          parser::IoControlSpec::CharExpr::Kind::Advance) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{
+                "ADVANCE specifier not allowed in DO CONCURRENT"_err_en_US});
+      }
+    }
+  }
+
+private:
+  bool ObjectIsCoarray() { return false; } // placeholder
+  bool EndTDeallocatesCoarray() { return false; } // placeholder
+
+  parser::CharBlock charBlock_;
+  parser::Messages &messages_;
+};
+
+class DoConcurrentCollection {
+public:
+  DoConcurrentCollection(parser::Messages &messages) : messages_{messages} {}
+  template<typename T> constexpr bool Pre(const T &) { return true; }
+  template<typename T> constexpr void Post(const T &) {}
+  bool Pre(const parser::ExecutionPart &executionPart) {
```

Comment:
This algorithm is similar to the one in labeled DO loop canonicalization and has similar issues with readability and correctness.  Let's resolve labeled DO loop canonicalization first and then it will be more clear what should be done here.

Reply:
I'll revisit. Thanks.